File: android/src/com/android/tools/idea/res/aar/Base128OutputStream.java
/*
 * Copyright (C) 2018 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.android.tools.idea.res.aar;

import java.io.BufferedOutputStream;
import java.io.IOException;
import java.io.OutputStream;
import java.nio.file.Files;
import java.nio.file.Path;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

/**
 * An output stream that uses the unsigned little endian base 128 (<a xref="https://en.wikipedia.org/wiki/LEB128">LEB128</a>)
 * variable-length encoding for integer values.
 * @see Base128InputStream
 */
class Base128OutputStream extends BufferedOutputStream {
  Base128OutputStream(@NotNull OutputStream stream) {
    super(stream);
  }

  Base128OutputStream(@NotNull Path file) throws IOException {
    super(Files.newOutputStream(file));
  }

  /**
   * Writes a 32-bit integer to the stream. Small positive integers take less space than larger ones:
   * <ul>
   *   <li>[0, 2^7) - 1 byte</li>
   *   <li>[2^7, 2^14) - 2 bytes</li>
   *   <li>[2^14, 2^21) - 3 bytes</li>
   *   <li>[2^21, 2^28) - 4 bytes</li>
   *   <li>[2^28, 2^31) - 5 bytes</li>
   *   <li>negative - 5 bytes</li>
   * </ul>
   * Avoid using this method for writing negative numbers since they take 5 bytes.
   *
   * @param value the value to write
   * @throws IOException if an I/O error occurs.
   */
  public final void writeInt(int value) throws IOException {
    do {
      int b = value & 0x7F;
      value >>>= 7;
      if (value != 0) {
        b |= 0x80;
      }
      super.write(b);
    } while (value != 0);
  }

  /**
   * Writes a 64-bit integer to the stream. Small positive integers take less space than larger ones:
   * <ul>
   *   <li>[0, 2^7) - 1 byte</li>
   *   <li>[2^7, 2^14) - 2 bytes</li>
   *   <li>[2^14, 2^21) - 3 bytes</li>
   *   <li>[2^21, 2^28) - 4 bytes</li>
   *   <li>[2^28, 2^35) - 5 bytes</li>
   *   <li>[2^35, 2^42) - 6 bytes</li>
   *   <li>[2^42, 2^49) - 7 bytes</li>
   *   <li>[2^49, 2^56) - 8 bytes</li>
   *   <li>[2^56, 2^63) - 9 bytes</li>
   *   <li>negative - 10 bytes</li>
   * </ul>
   * Avoid using this method for writing negative numbers since they take 10 bytes.
   *
   * @param value the value to write
   * @throws IOException if an I/O error occurs.
   */
  public final void writeLong(long value) throws IOException {
    do {
      int b = (int) value & 0x7F;
      value >>>= 7;
      if (value != 0) {
        b |= 0x80;
      }
      super.write(b);
    } while (value != 0);
  }

  /**
   * Write a String to the stream. The string is prefixed by its length + 1.
   * Each character is then written using the {@link #writeChar} method.
   *
   * @param str the string to write or null
   * @throws IOException if an I/O error occurs.
   */
  public final void writeString(@Nullable String str) throws IOException {
    if (str == null) {
      writeInt(0);
    }
    else {
      int len = str.length();
      writeInt(len + 1);
      for (int i = 0; i < len; i++) {
        writeChar(str.charAt(i));
      }
    }
  }

  /**
   * Writes a 16-bit integer to the stream. Small positive integers take less space than larger ones:
   * <ul>
   *   <li>[0, 2^7) - 1 byte</li>
   *   <li>[2^7, 2^14) - 2 bytes</li>
   *   <li>[2^14, 2^16) - 3 bytes</li>
   * </ul>
   *
   * @param value the value to write
   * @throws IOException if an I/O error occurs.
   */
  public final void writeChar(char value) throws IOException {
    writeInt(value & 0xFFFF);
  }

  /**
   * Writes a byte value to the stream.
   *
   * @param value the value to write
   * @throws IOException if an I/O error occurs.
   */
  public final void writeByte(byte value) throws IOException {
    super.write(value);
  }

  /**
   * Writes a boolean value to the stream.
   *
   * @param value the value to write
   * @throws IOException if an I/O error occurs.
   */
  public final void writeBoolean(boolean value) throws IOException {
    writeInt(value ? 1 : 0);
  }

  /**
   * @deprecated Use {@link #writeByte(byte)} or {@link #writeInt(int)} instead.
   * @throws UnsupportedOperationException when called
   */
  @Deprecated
  @Override
  public final void write(int b) {
    throw new UnsupportedOperationException(
        "This method is disabled to prevent unintended accidental use. Please use writeByte or writeInt instead.");
  }
}

File: streaming/screen-sharing-agent/app/src/main/cpp/base128_output_stream.cc
/*
 * Copyright (C) 2022 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "base128_output_stream.h"

#include <unistd.h>
#include <sys/socket.h>

#include <cassert>
#include <memory>

#include "io_exception.h"
#include "log.h"

namespace screensharing {

using namespace std;

Base128OutputStream::Base128OutputStream(SocketWriter&& writer, size_t buffer_size)
    : writer_(writer),
      buffer_(new uint8_t[buffer_size]),
      buffer_capacity_(buffer_size),
      offset_(0) {
}

Base128OutputStream::~Base128OutputStream() {
  Close();
  delete[] buffer_;
}

void Base128OutputStream::Close() {
  try {
    Flush();
  } catch (const IoException& e) {
    Log::E("Unable to flush Base128OutputStream");
  }
  shutdown(writer_.socket_fd(), SHUT_WR);
}

void Base128OutputStream::Flush() {
  if (offset_ > 0) {
    auto res = writer_.Write(buffer_, offset_, /*timeout_micros=*/10000000);
    if (res == SocketWriter::Result::DISCONNECTED) {
      throw EndOfFile();
    }
    if (res == SocketWriter::Result::TIMEOUT) {
      throw IoException();
    }
    offset_ = 0;
  }
}

void Base128OutputStream::WriteByte(uint8_t byte) {
  if (offset_ == buffer_capacity_) {
    Flush();
  }
  buffer_[offset_++] = byte;
}

void Base128OutputStream::WriteBytes(const string& bytes) {
  WriteInt32(bytes.size());
  for (auto b : bytes) {
    WriteByte(b);
  }
}

void Base128OutputStream::WriteUInt16(uint16_t value) {
  do {
    uint8_t b = value & 0x7F;
    value >>= 7;
    if (value != 0) {
      b |= 0x80;
    }
    WriteByte(b);
  } while (value != 0);
}

void Base128OutputStream::WriteUInt32(uint32_t value) {
  do {
    uint8_t b = value & 0x7F;
    value >>= 7;
    if (value != 0) {
      b |= 0x80;
    }
    WriteByte(b);
  } while (value != 0);
}

void Base128OutputStream::WriteUInt64(uint64_t value) {
  do {
    uint8_t b = value & 0x7F;
    value >>= 7;
    if (value != 0) {
      b |= 0x80;
    }
    WriteByte(b);
  } while (value != 0);
}

void Base128OutputStream::WriteBool(bool value) {
  WriteByte(value ? 1 : 0);
}

}  // namespace screensharing
